Migrate Dockerfiles from external repository

radusuciu opened the pull request:
## Description

This PR migrates the Dockerfiles currently housed in the OpenMS/dockerfiles as initially outlined in https://github.com/OpenMS/dockerfiles/issues/28. At the same time I'd like where possible optimize the images and related workflows to achieve minimal image size and build times

Some notes, outstanding questions, and issues. This is basically what I want to finish before this is ready for review/testing (though you're welcome to test and note issues ahead of time).

- [x] add a runtime stage which just copies all of the executables and only installs the bare minimum runtime libraries
- [x] fix use of `OPENMS_BRANCH` and `OPENMS_TAG`
- [x] ~~pyopenms target isn't building due to missing `python-dev` package. I copied this straight from the original dockerfile so I assume that that must've been broken as well~~
- [x] adjust software `LABEL` for each stage
- [x] fix related workflows, including singularity builds
- [x] use GHA cache
- [x] add `.dockerignore`


## Checklist
- [x] Make sure that you are listed in the AUTHORS file
- [x] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

radusuciu committed:
migrate OpenMS Dockerfile to main repository

radusuciu committed:
base docker images on external contrib image

radusuciu commented:
Two questions that haven't been addressed in the previous PRs/discussions:

1. Is the addition of a top-level `dockerfiles` folder okay?
2. Do we want to bring in the KNIME related dockerfiles as well?

@jpfeuffer, @timosachsenberg

radusuciu committed:
use separate build argument for OpenMS version

this is used in used for two things:
- selecting the base contrib image version
- selecting the branch from the OpenMS repository

radusuciu committed:
add stage-specific LABEL metadata

radusuciu commented:
So I've made good progress on most of the items outlined so far but the Singularity conversion isn't as simple as I was hoping. Though Singularity also has this concept, and `spython` converts it mostly fine (mostly because it doesn't do a good job of propagating `ARG` defaults between stages), I don't think there's an equivalent to `docker build --target STAGE`. Recent versions of Singularity have better support for running docker images directly without having to re-build but I think this requires the user to use the `--oci` option when running the container, which is not ideal. I have a few more ideas to try but please let me know if you have any thoughts!

radusuciu committed:
fix stage labels, disambiguate version and base image build args, and update workflow accordingly

radusuciu commented:
Update on the singularity issue mentioned above: instead of converting the `Dockerfile` to a Singularity `.def`, I'm attempting to convert the already built docker image directly: `sudo singularity build library.sif docker://ghcr.io/${{ needs.deploy-docker.outputs.downcase_repo }}-library:${{ needs.deploy-docker.outputs.tag_name }}` and have adjusted the workflows accordingly so that the singularity build job depends on the docker build job (and re-uses some metadata where possible).

I am not 100% sure if this is equivalent - this needs some testing. If this does not work then I see two choices:
1. Abandon the docker multi-stage build. I would prefer to avoid this as this is a perfect application of a multi-stage build.
2. Use `spython` to select the stages before converting to `.def`. I've tested this and it works -- the more annoying issue is that the way I'm using `ARG` in the new multi-stage `Dockerfile` will require some additional patching. This was already done in the existing workflow with `sed` so it's not too big of a blocker, but I'd prefer to avoid this if possible.

jpfeuffer commented:
There is no way to have build args or environment vars for singularity. And yes apparently no way to run a single stage.

I'm a bit hesitant of converting the ready images. It was a bit of the reason why we build them here. It sometimes failed. Not sure if it was because of on-the-fly conversions at the user and their environments but could also be due to incompatibilitied in the docker layers. I'm not sure.. might work. We could try..

An alternative could be to have a "from-scratch" singularity def file template that can be configured with jinja templates or something (for both args and stages). In this case, we should move the build steps to small scripts that will be mounted and executed, such that we do not have to maintain both files.

radusuciu commented:
Thank you for the context - I'll see if I can put something together, and if it doesn't work out maybe we can scale back the optimization ambitions and just do a direct migration, or you/others can take a stab at it.

Also, I should mention that I don't have much experience with singularity and don't understand deeply enough why converting an already built docker image might be flaky (and what the pitfalls are..), so I trust your experience of course. I did try and convert the current images though, and can verify that it does work or at least, I encounter no errors when doing the following:

```shell
docker run --rm -v $PWD:/go quay.io/singularity/singularity:v3.11.5 build executables.sif docker://ghcr.io/openms/openms-executables
docker run --rm -v $PWD:/go -it --privileged quay.io/singularity/singularity:v3.11.5 run executables.sif IsobaricAnalyzer --help
```
(yes, I'm running singularity in docker lol)

Is there a chance that newer versions of singularity are better in this regard? It does seem that docker/OCI compatibility has been a priority for them.

jpfeuffer commented:
I guess we can try with the converted containers and come up with a solution if they fail. Apparently, according to your external issue, they are pretty confident about the conversion process.

radusuciu commented:
> I guess we can try with the converted containers and come up with a solution if they fail. Apparently, according to your external issue, they are pretty confident about the conversion process.

I tried to learn a bit more about how it works, I think the layer flattening should be pretty robust. There's also an open issue with a kinda old PR for building directly SIF from Dockerfiles. Basically I think if we have issues it'd also be worth raising them with the Singularity project.

I would like to test the conversion process in a GitHub action though, in my fork. One thing that I can think of that would lead to sporadic failures would be worker diskspace.. I've hit this a few times when building larger containers and had to either slim them down or spread them across different jobs.

EDIT: The conversion/build works, and takes ~7 minutes: https://github.com/radusuciu/OpenMS/actions/runs/7587340359/job/20667464502

radusuciu committed:
add GHA cache to docker build workflow

radusuciu committed:
use more convervative default for NUM_BUILD_CORES build arg

NOTE: in github actions this is set to 2

jpfeuffer commented:
Yep I think the disk space and permissions on clusters is what made us build them here for other people so they don't have to do it on the fly. I guess if we can convert in a controlled environment, such as here it should really be fine. I'm very happy to merge. Happy to see that it works out on your fork as well.

radusuciu commented:
Alright, cool! I'm now focusing on seeing if there's any fat to trim from the images. For example, the main repository .git and THIRDPARTY submodules (note: everything needed for the container is already copied to `/thirdparty`) which total up to ~2GB in the current executables image:

```
$ docker run -it --rm ghcr.io/openms/openms-executables:3.1.0 du -sh /OpenMS/{THIRDPARTY,.git}/
1.1G    /OpenMS/THIRDPARTY/
970M    /OpenMS/.git/
```

That's out of a total of 5.3GB so that's a pretty big chunk:

```
$ docker image inspect -f "{{ .Size }}" ghcr.io/openms/openms-executables:3.1.0 | numfmt --to=si
5.3G
```

radusuciu commented:
Before PR is ready for review, I had a few questions for you (or any other maintainers):

1. Is there any test-suite we can run on the executables/tools alone? I'm asking because I've already played around with a minimal runtime stage which is just executables and dependencies from apt (ie. no `-dev` packages, no unnecessary build tools), and while I've confirmed that they run without errors, I would love if it I could run the same test suite that is run during build (this would also be a separate stage)
2. Are you currently doing anything with the pyOpenMS docker container: https://github.com/OpenMS/dockerfiles/blob/master/pyOpenMS/Dockerfile. If not, I will not migrate this, though I'm happy to contribute a separate PR later if there is a potential use.
3. Are you currently doing anything with the KNIME docker containers: https://github.com/OpenMS/dockerfiles/tree/master/KNIME. If not, I will not migrate this, though I'm happy to contribute a separate PR later if there is a potential use.
4. I forget what motivated me but I introduced an alternative tag for `openms-executables`: `openms-tools`. Do you have any opinion there?

timosachsenberg commented:
> 1. Is there any test-suite we can run on the executables/tools alone? I'm asking because I've already played around with a minimal runtime stage which is just executables and dependencies from apt (ie. no -dev packages, no unnecessary build tools), and while I've confirmed that they run without errors, I would love if it I could run the same test suite that is run during build (this would also be a separate stage)

hmm I think this is not easily done because installing tests is a bit non-standard for cmake projects.
The test project is created here https://github.com/OpenMS/OpenMS/blob/develop/src/tests/topp/CMakeLists.txt
And it seems to me it is not easy to pull that out of the project.

What others have done: they take one or few of the tests (preferably of a stable tool, maybe OpenMSInfo) and execute it in the container to check if loading of libraries etc. works. 

Edit: fixed sentence

jpfeuffer commented:
You can just copy the test executables and run them. Some of them will have hard coded paths to compare outputs to expected outputs. Those files must be copied over to the same path. Not sure if those were relative paths. Other tests will work out of the box.

You can also copy over all ctest related files and with Cmake 3.20 run `ctest --test-dir new_dir_on_target_machine`. Figuring out what you need might be non-trivial. We never tried this. It would be cool to have a Cmake install configuration that installs expected test files as well but I think it's a bit of work.

I wouldn't say our tests are non-standard at all.
I think testing a full test suite on a target (non-build) machine is not toooo common yet and not well supported in CMake in general (until maybe recently)

radusuciu commented:
Thanks for the responses, I think I've got it! I've reduced the tools/executables runtime to below 1GB.
```
$ docker image inspect -f "{{ .Size }}" tools-slim | numfmt --to=si
911M
```

The relevant stage:

```Dockerfile
FROM ubuntu:22.04 AS tools-slim
ARG OPENMS_LIBRARY_DIR
ARG OPENMS_LIBRARY_BUILD_DIR
ARG CONTRIB_BUILD_DIR
ARG OPENMS_LIBRARY_DIR

RUN apt-get update \
  && apt-get install -y --no-install-recommends --no-install-suggests \
    libqt5opengl5 \
    libsvm3 \
    libzip4 \
    zlib1g \
    libbz2-1.0 \
    libgomp1 \
    libqt5svg5 \
    libxerces-c3.2 \
    libboost-date-time1.74-dev  \
    libboost-iostreams1.74-dev \
    libboost-regex1.74-dev \
    libboost-math1.74-dev \
    libboost-random1.74-dev \
  && rm -rf /var/lib/apt/lists/*

COPY --from=tools ${CONTRIB_BUILD_DIR}/bin ${CONTRIB_BUILD_DIR}/bin
COPY --from=tools /thirdparty /thirdparty
COPY --from=tools ${OPENMS_LIBRARY_BUILD_DIR}/bin ${OPENMS_LIBRARY_BUILD_DIR}/bin
COPY --from=tools ${OPENMS_LIBRARY_BUILD_DIR}/lib ${OPENMS_LIBRARY_BUILD_DIR}/lib
COPY --from=tools ${OPENMS_LIBRARY_DIR}/share ${OPENMS_LIBRARY_DIR}/share
```

What's missing is the `PATH` env. Though I think it might be worthwhile to move things from the various build dirs to more standard system locations (this is where we might need to much about with `ctest` arguments). I'm testing in a separate stage based on `tools-slim` where I install cmake and the test files, and just got here:

```bash
root@US320024WSD:/openms-build# ctest -R TOPP_

[snipped stdout]

100% tests passed, 0 tests failed out of 1278
```

😃

I'll update PR today/tomorrow, but I think this is very promising and IMO a worthwhile optimization if it holds.

jpfeuffer commented:
Whaat? That's awesome!

radusuciu committed:
add a slimmer tools stage

a few other modifications:
- added dockerignore file
- clone thirdparty repo instead of using submodule since I've added .git to the dockergignore

radusuciu commented:
Okay, so the slimmed tools image is at 1.3GB, I accidently the `/thirdparty` before, but still, pretty good. I think we're fairly close to the minimum now - take a look at the [`dive`](https://github.com/wagoodman/dive) layer summary:

```
│ Layers ├─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Cmp   Size  Command                                                                                                                
     78 MB  FROM 9ee4303fd391873                                                                                                   
    553 MB  RUN |5 OPENMS_LIBRARY_DIR=/OpenMS OPENMS_LIBRARY_BUILD_DIR=/openms-build CONTRIB_BUILD_DIR=/contrib-build OPENMS_LIBRAR
    9.4 MB  COPY /contrib-build/bin /contrib-build/bin # buildkit                                                                  
    370 MB  COPY /thirdparty /thirdparty # buildkit                                                                                
     19 MB  COPY /openms-build/bin /openms-build/bin # buildkit                                                                    
     49 MB  COPY /openms-build/lib /openms-build/lib # buildkit                                                                    
    202 MB  COPY /OpenMS/share /OpenMS/share # buildkit                                                                            
```

Notes on the above:
- the runtime deps (553MB) are the biggest contributor, and I don't think these can be trimmed much more (note the full command is cut off but you can see it in the Dockerfile)
- the base system is only 78MB. Earlier I suggested trying out debian-slim, but that's only a few MB smaller, so not sure that's worth it at this point
- `/OpenMS/share/OpenMS/examples` is 136MB, so maybe worth ditching

All binaries/files are in the original build folders still, I'll see if I can tidy that up, but that's almost an aesthetic concern.

radusuciu committed:
fix missing checkout action

radusuciu committed:
fix missing Dockerfile

jpfeuffer commented on .github/workflows/containerdeploy.yml:
```diff
@@ -30,144 +33,99 @@ jobs:
           build-args: |
-            OPENMS_TAG=latest
-            OPENMS_BRANCH=${{ steps.extract_branch.outputs.branch }}
+            NUM_BUILD_CORES=2
```
I think GitHub recently upgraded their free runners. You might try 4.

radusuciu replied:
Nice! You're right: https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories

jpfeuffer commented on dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,170 @@
+    rm -rf THIRDPARTY
+EOF
+
+ENV PATH="${THIRDPARTY_DIR}/LuciPHOr2:${THIRDPARTY_DIR}/MSGFPlus:${THIRDPARTY_DIR}/Sirius:${THIRDPARTY_DIR}/ThermoRawFileParser:${THIRDPARTY_DIR}/Comet:${THIRDPARTY_DIR}/Fido:${THIRDPARTY_DIR}/MaRaCluster:${THIRDPARTY_DIR}/Percolator:${THIRDPARTY_DIR}/SpectraST:${THIRDPARTY_DIR}/XTandem:${THIRDPARTY_DIR}/Sage:${PATH}"
```
Not sure but very recently we switched our support policy for Sirius. We only read their workspaces now and do not provide adapters anymore due to the size of their package. You might be able to remove it.

radusuciu replied:
Will you be removing it from the thirdparty repo?

jpfeuffer replied:
Yes I think so. Maybe we did already @axelwalter ?

axelwalter replied:
The changes regarding Sirius tools are in PR #7234. Once merged, I will remove Sirius from THIRDPARTY repository.

jpfeuffer commented on dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,170 @@
+
+RUN apt-get update \
+  && apt-get install -y --no-install-recommends --no-install-suggests \
+    libqt5opengl5 \
```
We might think about a distinction between tools and gui-tools.
GUI tools will install a ton of other stuff.

radusuciu replied:
Good point. Is it just a matter of excluding dependencies for GUI tools in the `tools` and including them in `gui-tools`? Or are there some GUI only tools that we shouldn't build at all?

jpfeuffer replied:
You can re-use an existing OpenMS build with "-DWITH_GUI=OFF", install additional dependencies such as qt5 gui components and then re-configure CMake with "-DWITH_GUI=ON". Then one can use `make OpenMS_GUI` to build the GUI tools only.
This could be just an additional layer on top of the tools image.

radusuciu replied:
Alright, so I did this and the difference is 7.7 MB for the library image and 41 MB for the tools image (they're just under 900 MB). The `Dockerfile` (https://github.com/radusuciu/OpenMS/commit/b9c4defc495066ffd12133c247a0a73cc774ac06) is a bit more complex though. I'm not sure it's worth the tradeoff especially when we add a `-thirdparty` variant which will make a much bigger difference.

It's possible I'm doing something wrong though. I did as you said:
- for the `library-build` stage I configure with `-DWITH_GUI=OFF`
- for the `library-build-gui` stage reconfigured with  `-DWITH_GUI=ON`, then `make OpenMS_GUI`
- I only install `libqt5opengl5` in the `library` stage
- the `tools` and `tools-gui` images are mostly the same -- I still `make TOPP` in the build stage. Then, I copy the library from the corresponding build stage, and I install `openjdk-17-jre` for the GUI and `openjdk-17-jre-headless` for the regular image.

IMO, unless I've missed something I'd rather aim for a `-thirdparty` variant, merge, and then revisit the GUI issue.

jpfeuffer commented on dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,170 @@
+    libboost-iostreams1.74-dev \
+    libboost-regex1.74-dev \
+    libboost-math1.74-dev \
+    libboost-random1.74-dev \
```
I'm kind of missing an ILP solver here. CoinOR (preferably due to license) or GLPK.
I wonder how this successfully configures.

jpfeuffer replied:
Also for full functionality you might want `mono/.NET` for ThermoRawFileParser. But it's  a beast so I am not sure.

radusuciu replied:
Hmm, good point, I will check out the CoinOR issue -- is there a test that should catch this?

Regarding `mono` or `.NET`, I lean towards not including it unless some OpenMS component depends on it directly, or maybe if you included it in your THIRDPARTY repository.

jpfeuffer replied:
Regarding CoinOR: LPWrapper_test or something should usually fail. Actually the CMake configuration stage that looks for dependencies should already fail. So I assume it is finding it somewhere. Are you currently testing with the contrib image as a base image maybe? In that case none of the apt dependencies are needed of course. Contrib and apt dependencies (except for Qt which is huge :( ) are mutually exclusive.

Regarding mono: We could actually also distinguish between openms-tools and openms-thirdparty as we do on conda.
Where only openms-thirdparty contains THIRDPARTY and its dependencies.
I just mean: If we do not install mono for a lightweight image, it does not make sense to pull the ThermoRawFileParser libs and bins from our THIRDPARTY repo, as they will be unusable on the image in that state.

radusuciu replied:
Regarding `CoinOR`, I think we're covered because in the current `contrib` image, it's being built and patched separately and included in this image like so: `COPY --from=tools-build ${CONTRIB_BUILD_DIR}/bin ${CONTRIB_BUILD_DIR}/bin`. I think the consensus on that is that we should move to installing it from apt, in which case, yes any related tests should fail. However, note that `LPWrapper_test` is not run - the test command is `ctest -R TOPP_`. Does it make sense to run the whole test suite for this image?

(I haven't forgotten about the `mono` related bits, let's revisit after we're clear on `CoinOR`)

jpfeuffer replied:
Yes correct. It would be cool if during the transition phase we could choose between a build that copies from the contrib container or one that downloads from apt.
But mid-term apt is easier to maintain.

jpfeuffer replied:
Maybe for openms-lib we can run class tests and for openms-tools the TOPP tests.
But for now TOPP tool tests should cover most of it.

radusuciu replied:
> It would be cool if during the transition phase we could choose between a build that copies from the contrib container or one that downloads from apt.

If we merge https://github.com/OpenMS/contrib/pull/136 first, then we can just go with apt here, no? I don't think anything should break. I'd rather avoid that if you agree.

> Maybe for openms-lib we can run class tests and for openms-tools the TOPP tests. But for now TOPP tool tests should cover most of it.

Okay, I think that running the class tests for the library makes sense. You have the tests, we might as well make use of them :)

jpfeuffer commented on dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,170 @@
+ARG THIRDPARTY_DIR
+ENV MAKEFLAGS="${MAKEFLAGS}"
+
+RUN apt-get update && apt-get install -y --no-install-recommends --no-install-suggests openjdk-17-jdk
```
JRE might be enough. Did you try?
It's for msgf and luciphor.

radusuciu replied:
I did not try, but good suggestion, thanks!

radusuciu replied:
Just to be clear, is this a build or runtime dependency?

jpfeuffer replied:
I think only runtime/testing with thirdparty.

jpfeuffer reviewed (commented):
Added some quick comments

jpfeuffer commented:
Regarding runtime binaries.
If you do not `make install` the binaries and libs won't be stripped. Another path to explore. 
Make install will install out of the build directory though. Which could mess up your nice tests.
Maybe run your own strip.
Pros: Smaller binaries
Cons: probably no function info in stack traces etc.

radusuciu committed:
go back to cloning inside build

copying from the CWD gave me a bunch of issues while testing locally (eg. wrong context, accidental cache invalidation etc.) that I think it's better to just `git clone`.

TODO: ensure we're not unnecessarily including the .git folder now

radusuciu committed:
run `update-ca-certificates` in same command as the package installs

radusuciu committed:
remove accidental newline at beginning of dockerignore

radusuciu committed:
fix cmake_prefix_path setting

single quotes meant that we were literally adding ${CONTRIB_BUILD_DIR} to the prefix path

radusuciu committed:
make use of THIRDPARTY_DIR build arg instead of hardcoding

maybe I've gone overkill on the build args here but I think it helps reduce the chance of errors when dealing with multiple stages/images and makes it easier to move things around

radusuciu committed:
use 4 build cores for container deploy action

radusuciu committed:
remove Sirius from docker image path

Sirius adapters are in the process of being removed in #7234

radusuciu committed:
move install in docker image from /OpenMS to /opt/OpenMS

radusuciu committed:
install JRE for docker tools runtime dependency

radusuciu committed:
add a library-build stage to minimize the size of the library container

also includes some fixes to ensure that the OpenMS.git does not come along for the ride in the final runtime images

radusuciu committed:
test both tools and library runtime docker images

radusuciu committed:
actually copy in the thirdparty deps..

radusuciu committed:
ensure we fail out of any multiline scripts

radusuciu committed:
remove -o pipefail since docker RUN commands are run in /bin/sh which does not support this options

radusuciu committed:
add a contrib stage as an alternative base to that hosted in OpenMS/contrib

NOTE: this is mostly for testing and local builds and it is not the default.

radusuciu committed:
fix docker contrib stage

we need a bin subdirectory.

radusuciu committed:
move cmake install after other packages

radusuciu committed:
install coinor runtime dependency

radusuciu committed:
docker: add mono runtime dependency for ThermoRawFileParser

radusuciu committed:
docker: fix DEBIAN_FRONTEND setting

radusuciu committed:
docker: fix to apply DEBIAN_FRONTEND envvar properly

radusuciu commented:
Alright, I've made a few changes:
- added `mono` to the runtime tools image. This is indeed pretty heavyweight. We could work around this by building ThermoRawFileParser with `dotnet` instead, and bundling that into the thirdparty repo. Not sure if it's worth the effort.
- optimized the library image using a library specific build stage
- added a contrib stage for testing. The image built from OpenMS/contrib is still the default, as discussed.
- I'm now running all of the tests instead of just `TOPP_` in a single stage. To do this, I run `make all` in a separate stage. It's possible that we can be more specific..

We're up to 1.8G now for the tools image (mostly JRE/mono). Still not a bad improvement IMO and there's room to improve thanks to advice from @jpfeuffer, summarized below.

Here's what we can still do:
- [x] Create separate thirdparty image that includes all thirdparty tools (and their dependencies) in addition to OpenMS executables
- [x] ~~Create separate GUI image -- though if we want thirdparty, does this mean we will have 4 variants of the tools?~~
- [x] running `make install` to strip binaries -- though there are some tradeoffs to consider here

Other than that things are looking good. 2 tests are failing, which I will investigate later:

```
The following tests FAILED:
  619 - TOPPView_test (Failed)
  620 - TSGDialog_test (Failed)
```

radusuciu committed:
docker: separate tools and tools+thirdparty images

radusuciu committed:
docker: switch back to copying source from context instead of using git clone

Rationale for this is to make it possible to build this in forks or locally

radusuciu committed:
docker: add readme describing images

can do more work here

radusuciu committed:
docker: add dependency for running tests

radusuciu committed:
docker: create the contrib build dir if it doesn't exist

radusuciu commented:
Alright, I'm approaching being ready for final review. I've sorted out those last two tests, and now just trying to see if I can get `make install` and still run the tests with runtime dependencies.

radusuciu committed:
docker: use `make install` and refactor to use single build stage

also:
- renamed some docker ARGs.. we don't need to prefix everything with OPENMS_

radusuciu committed:
Merge branch 'OpenMS:develop' into dockerfiles

radusuciu committed:
docker: tweak workflow to ensure we have a .git folder during build so we can submodule update

radusuciu committed:
docker: keep git dir when building docker images

radusuciu committed:
docker: remove gha cache scope arg

this was overly specific. we might want to have a scope eventually but I don't think caching was really working before. I'm finding that information on this is a bit sparse, perhaps this makes sense given that this is labeled "experimental"..

radusuciu committed:
docker: rename OPENMS_VERSION ARG to OPENMS_VERSION_TAG for clarity

radusuciu committed:
docker: set contrib image tag to latest unless

this is a complication of having the contrib dockerfile built in a separate repository. During a release, we build it here and the tag is consistent between the contrib image and the other images we build. For non-release pushes, we have to make sure that the tag we use to pull that image is "latest" or else this action will fail on other branches that may not exist in the contrib repository.

radusuciu committed:
docker: run tests as first step of workflow

radusuciu committed:
docker: use consistent case for "AS"

radusuciu committed:
docker: remove currently unused pyopenms stage

radusuciu commented:
Alright, so `make install/strip` offers some modest savings in final image size and I think it's a generally cleaner approach. I think the test stage is still valuable, though I want to be clear on what is being tested here. The test stage is based on the tools+thirdparty image which contains runtime dependencies installed on top of a fresh Ubuntu image. Then, we copy in the build and source directories, and run the tests. So we're not testing the installed/stripped binaries and library, but instead we're testing that the output of the build is functional with a different set of dependencies. I think this is reasonable and addresses my original concern, but please let me know if you see issues here.

I am hitting two failures on tests that I haven't been able to figure out and would appreciate any pointers. See this workflow run in my fork: https://github.com/radusuciu/OpenMS/actions/runs/7820217460

<details><summary>Tests should pass now, so this is mostly irrelevant..</summary>

```
The following tests FAILED:
	287 - InternalCalibration_test (Failed)
	607 - TOPPView_test (Failed)
```

I suspect they're related to the location of the share directory, and where it gets installed but I haven't managed to sort it out even though I've placed this directory everywhere I could think of and I've set the `OPENMS_DATA_PATH` envvar as well. Note that I also made use of `-DCMAKE_INSTALL_PREFIX` (maybe related: https://github.com/OpenMS/OpenMS/issues/1479).
 
```
1023/2407 Test  #287: InternalCalibration_test ...................................***Failed    0.18 sec
checking InternalCalibration() ... 
: passed

checking ~InternalCalibration() ... : passed
Warning: no subtests performed in '~InternalCalibration()' (line 44)!

checking Size fillCalibrants(const PeakMap exp, const std::vector<InternalCalibration::LockMass>& ref_masses, double tol_ppm, bool lock_require_mono, bool lock_require_iso, CalibrationData& failed_lock_masses, bool verbose = true) ... Lock masses found across viable spectra:
  0 [of 3] lock masses: 108x
  3 [of 3] lock masses: 21x


: passed

Error: Caught unexpected OpenMS exception of type 'FileNotFound' thrown in line 417 of file '/tmp/OpenMS/src/openms/source/SYSTEM/File.cpp' in function 'static OpenMS::String OpenMS::File::find(const OpenMS::String&, OpenMS::StringList)' - Message: the file './examples/BSA/BSA1_OMSSA.idXML' could not be found
checking (created temporary files)...
: passed

Output of successful tests were suppressed. Set the environment variable 'OPENMS_TEST_VERBOSE=True' to enable them.
FAILED
Failed lines: 
```

Not sure what's going on here.. Is the path correct in the test? Shouldn't it point to the share directory?

```
1493/2407 Test  #607: TOPPView_test ..............................................***Failed    1.21 sec
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
********* Start testing of OpenMS::TestTOPPView *********
Config: Using QtTest library 5.15.3, Qt 5.15.3 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 11.3.0), ubuntu 22.04
PASS   : OpenMS::TestTOPPView::initTestCase()
FAIL!  : OpenMS::TestTOPPView::testGui() Compared values are not the same
   Actual   (tv.tab_bar_.tabText(tv.tab_bar_.currentIndex())): ""
   Expected (QString("peakpicker_tutorial_1 (1D)"))          : "peakpicker_tutorial_1 (1D)"
   Loc: [/tmp/OpenMS/src/tests/class_tests/openms_gui/source/GUI/TOPPView_test.cpp(109)]
PASS   : OpenMS::TestTOPPView::cleanupTestCase()
Totals: 2 passed, 1 failed, 0 skipped, 0 blacklisted, 1154ms
********* Finished testing of OpenMS::TestTOPPView *********
```

For this failure, I guess I'm not installing all of the runtime dependencies for the GUI. Perhaps we should just configure with `-DWITH_GUI=OFF`? I imagine that running these images with a GUI isn't a common use-case..
</details>

radusuciu committed:
docker: fix tests not passing due to missing examples

radusuciu commented:
Alright.. tests should pass now: https://github.com/radusuciu/OpenMS/actions/runs/7805864399. The issue was that those two tests were looking for files in `share/examples` which were not there.

`make install` with the install prefix set to `/opt/OpenMS` places the share directory in `/opt/OpenMS/share/OpenMS`, but the examples end up in `/opt/OpenMS/share/examples/examples`, whereas in the source they are not nested in two `examples` folders. By copying the share directory from the source, as opposed to the install directory, we avoid this double-nesting problem.

Is there a bug here: https://github.com/OpenMS/OpenMS/blob/1c07f3cb3a70e5c4796adfcaee3268e782351aac/src/openms/CMakeLists.txt#L177

should perhaps be:

```
install_directory(${OPENMS_HOST_DIRECTORY}/share/OpenMS/examples ${INSTALL_SHARE_DIR} examples)
```

radusuciu committed:
Merge remote-tracking branch 'upstream/develop' into dockerfiles

radusuciu committed:
docker: copy share from install directory now that #7337 has been merged

radusuciu committed:
docker: update readme and add changelog entry

radusuciu committed:
docker: remove unnecessary arg

radusuciu committed:
docker: whitespace differences

radusuciu committed:
docker: do not push test image

it is sufficient for caching purposes to load it

timosachsenberg committed:
Merge branch 'develop' into dockerfiles

radusuciu committed:
docker: free up space on runner by deleting unneccessary tools

runners have varying amounts of free space, and these images can be quite big and sometimes juuuust fail to build if the runner happens to have less free space.

poshul committed:
Merge branch 'develop' into dockerfiles

radusuciu commented:
Thanks for keeping the PR up to date. It's pretty much ready for merge but I would recommend a final review. I've been using a similar image that I build here https://github.com/radusuciu/docker-openms/ without problems.

There are two things I'd call your attention to:
1. I added a new tag `openms-tools`, but kept the original as well `openms-executables`.. maybe you prefer to not change this or to not have two.
2. I have not tested this latest version with a release tag. In this case, the workflow builds the contrib image. I can test this to make sure there are no issues.

timosachsenberg commented:
looks very good to me in general.
One part where I am a bit lost is the contrib.
In the Dockerfile we use system packages and in the action we use the submodule.
Is there a reason why this is done this way e.g. to preserve both ways to provide dependencies?

radusuciu commented:
> looks very good to me in general. One part where I am a bit lost is the contrib. In the Dockerfile we use system packages and in the action we use the submodule. Is there a reason why this is done this way e.g. to preserve both ways to provide dependencies?

Hey, sorry for delay, I just got back from a trip.

This was mainly for easy testing prior to https://github.com/OpenMS/contrib/pull/136 being merged. It might still be helpful to have as a stage in the same Dockerfile as opposed to a separate Dockerfile in the sub-module when how changes to the contrib packgages affect builds of the images here.. but this also introduces the risk of some confusion for future maintenance.

I don't feel super-strongly, but I'm inclined to say let's remove it.

timosachsenberg commented:
@jpfeuffer what is your call here?

jpfeuffer commented:
If you want to avoid contrib containers you can also build remaining dependencies in the corresponding action (CI, pyopenms, this docker action)
Even better, add it to cmake and build automatically if not found, as we do for json and SQLite.
You can also cache the deps but our caches are already at the limit due to compiler caches

jpfeuffer commented:
Ah I misremembered the end of the thread here. I think you can remove.
The inability to get the full contrib on manylinux does not stop us from building the Ubuntu containers here.

(For which we do not need any contrib base containers anymore)

radusuciu committed:
remove contrib stage (was only used in testing)

radusuciu committed:
Merge branch 'dockerfiles' of https://github.com/radusuciu/OpenMS into dockerfiles

radusuciu commented:
I've moved the unused contrib stage from the `Dockerfile`. Some of the points I raised in the comment above might still warrant comment -- let me know or we can merge and deal with them separately.

Thanks!

poshul committed:
Merge branch 'develop' into dockerfiles

poshul commented:
Is this ready to merge? I'm trying to set up devcontainer image that we can use for development and came across this (which is going to be used as a basis for my setup)

radusuciu commented:
Quoting myself

> There are two things I'd call your attention to:
> 1. I added a new tag `openms-tools`, but kept the original as well `openms-executables`.. maybe you prefer to not change this or to not have two.

This is minor, and well covered in the README I included in the containing directory: https://github.com/OpenMS/OpenMS/blob/54a90800c60e9ffd9a1ec250671480277a25bb74/dockerfiles/README.md

> 2. I have not tested this latest version with a release tag. In this case, the workflow builds the contrib image. I can test this to make sure there are no issues.

I just tagged a fake release, let's see if it runs https://github.com/radusuciu/OpenMS/actions/runs/10724213483

radusuciu commented:
Not necessarily related to the PR but looks like we failed two tests

```
#37 31.24           Start 2122: TOPP_THERMORAWFILEPARSER_1
#37 31.66  652/2353 Test #2122: TOPP_THERMORAWFILEPARSER_1 .................................***Failed    0.43 sec
#37 31.67 RawFileReader reading tool. Copyright 2016 by Thermo Fisher Scientific, Inc. All rights reserved
#37 31.67 Standard output: Running: mono /opt/OpenMS/thirdparty/ThermoRawFileParser/ThermoRawFileParser.exe -i=/tmp/OpenMS/src/tests/topp/THIRDPARTY/ginkgotoxin-ms-switching.raw --output_file=ginkgotoxin-ms-switching_out_tmp.mzML -f=2 -e
#37 31.67 
#37 31.67 Standard error: Unexpected extra arguments
#37 31.67 Error - usage is:
#37 31.67   -h, --help                 Prints out the options.
#37 31.67       --version              Prints out the version of the executable.
#37 31.67   -i, --input=VALUE          The raw file input (Required). Specify this or an
#37 31.67                                input directory -d.
#37 31.67   -d, --input_directory=VALUE
#37 31.67                              The directory containing the raw files (Required).
#37 31.67                                Specify this or an input raw file -i.
#37 31.67   -b, --output=VALUE         The output file. Specify this or an output
#37 31.67                                directory -o. Specifying neither writes to the
#37 31.67                                input directory.
#37 31.67   -o, --output_directory=VALUE
#37 31.67                              The output directory. Specify this or an output
#37 31.67                                file -b. Specifying neither writes to the input
#37 31.67                                directory.
#37 31.67   -s, --stdout               Write to standard output. Cannot be combined with
#37 31.67                                file or directory output. Implies silent logging,
#37 31.67                                 i.e. logging level 0
#37 31.67   -f, --format=VALUE         The spectra output format: 0 for MGF, 1 for mzML,
#37 31.67                                2 for indexed mzML, 3 for Parquet, 4 for None (
#37 31.67                                no output); both numeric and text (case
#37 31.67                                insensitive) value recognized. Defaults to
#37 31.67                                indexed mzML if no format is specified.
#37 31.67   -m, --metadata=VALUE       The metadata output format: 0 for JSON, 1 for TXT,
#37 31.67                                2 for None (no output); both numeric and text (
#37 31.67                                case insensitive) value recognized. Defaults to
#37 31.67                                None
#37 31.67   -c, --metadata_output_file=VALUE
#37 31.67                              The metadata output file. By default the metadata
#37 31.67                                file is written to the output directory.
#37 31.67   -g, --gzip                 GZip the output file.
#37 31.67   -p, --noPeakPicking[=VALUE]
#37 31.67                              Don't use the peak picking provided by the native
#37 31.67                                Thermo library. By default peak picking is
#37 31.67                                enabled. Optional argument allows disabling peak
#37 31.67                                peaking only for selected MS levels and should
#37 31.67                                be a comma-separated list of integers (1,2,3)
#37 31.67                                and/or intervals (1-3), open-end intervals (1-)
#37 31.67                                are allowed
#37 31.67   -z, --noZlibCompression    Don't use zlib compression for the m/z ratios and
#37 31.67                                intensities. By default zlib compression is
#37 31.67                                enabled.
#37 31.67   -a, --allDetectors         Extract additional detector data: UV/PDA etc
#37 31.67   -l, --logging=VALUE        Optional logging level: 0 for silent, 1 for
#37 31.67                                verbose, 2 for default, 3 for warning, 4 for
#37 31.67                                error; both numeric and text (case insensitive)
#37 31.67                                value recognized.
#37 31.67   -e, --ignoreInstrumentErrors
#37 31.67                              Ignore missing properties by the instrument.
#37 31.67   -x, --excludeExceptionData Exclude reference and exception data
#37 31.67   -L, --msLevel=VALUE        Select MS levels (MS1, MS2, etc) included in the
#37 31.67                                output, should be a comma-separated list of
#37 31.67                                integers (1,2,3) and/or intervals (1-3), open-
#37 31.67                                end intervals (1-) are allowed
#37 31.67   -P, --mgfPrecursor         Include precursor scan number in MGF file TITLE
#37 31.67   -N, --noiseData            Include noise data in mzML output
#37 31.67   -w, --warningsAreErrors    Return non-zero exit code for warnings; default
#37 31.67                                only for errors
#37 31.67   -u, --s3_url[=VALUE]       Optional property to write directly the data into
#37 31.67                                S3 Storage.
#37 31.67   -k, --s3_accesskeyid[=VALUE]
#37 31.67                              Optional key for the S3 bucket to write the file
#37 31.67                                output.
#37 31.67   -t, --s3_secretaccesskey[=VALUE]
#37 31.67                              Optional key for the S3 bucket to write the file
#37 31.67                                output.
#37 31.67   -n, --s3_bucketName[=VALUE]
#37 31.67                              S3 bucket name
#37 31.67 Process 'mono' did not finish successfully (exit code: 255). Please check the log.
```

and

```
#37 97.67 2195/2353 Test #2123: TOPP_THERMORAWFILEPARSER_1_out .............................***Failed    0.02 sec
#37 97.67 Warning: Parameters file version (3.2.0) does not match the version of this tool (3.2.0-pre-HEAD-2024-09-05).
#37 97.67 Your current parameters are still valid, but there might be new valid values or even new parameters. Upgrading the INI might be useful.
#37 97.67 Cannot read input file given from parameter '-in1'!
#37 97.67 Error: File not found (the file 'ginkgotoxin-ms-switching_out_tmp.mzML' could not be found)
#37 97.67 
```

poshul committed:
Merge branch 'develop' into dockerfiles

poshul commented:
So the error is due to a change in the command line params to ThermoRawFileParser we just need to make sure to that the test cmake file matches the correct version of the executable.

timosachsenberg commented:
@poshul what changed?
- Needs to be fixed in cmake scripts
- This probably needs to be fixed in FileConverter as well.

Deployment fails currently

timosachsenberg commented:
fixed in https://github.com/OpenMS/OpenMS/pull/7673
